src/python/pants/pantsd/pants_daemon.py

```diff
@@ -332,7 +332,7 @@ def _run_services(self, services):
 
     # Once all services are started, write our pid.
     self.write_pid()
-    self.write_metadata_by_name('pantsd', self.FINGERPRINT_KEY, self.options_fingerprint)
+    self.write_metadata_by_name('pantsd', self.FINGERPRINT_KEY, self.options_fingerprint.decode('utf-8'))
```

Comment:
This was wild to debug. 

> Pantsd was throwing this exception in integration tests:
Exception message: abruptly lost active connection to pantsd runner: NailgunError(u"Problem talking to nailgun server (address: 127.0.0.1:52211): error(54, 'Connection reset by peer')",)
>
> It turns out this had nothing to do with Nailgun Server. Instead, the issue was `safe_file_dump` silently failing to write metadata for pantsd. The open backport should theoretically have thrown a TypeError with this, but it didn't for some reason. I believe Py3 would have thrown an error, although cannot test due to the Rust ImportError.
ecda20b  

In the process, I realized `hasher.hexdigest()` has inconsistent behavior between Py2 and Py3. In Py2, it returns bytes. In Py3, it returns unicode. As a separate PR, I'm going to change our usage of sha1() so that we always return unicode. I think it will catch some issues for us, similar to this `open` backport project.